### PR TITLE
Add actual meaning error to string

### PR DESCRIPTION
**Pull request by DominicWatts**
Adding meaningful error to string - I've seen a few different methods of generating this error online and I think it will help everyone if you see the actual error - not what's currently displayed

### Description (*)

Add meaning eror to end of string

### Fixed Issues (if relevant)

Lots of reports online regarding this message

### Manual testing scenarios (*)

In my case I provided duplicate category IDs in an array to assignProductToCategories method as part of linking products to categories process.  Once I revealed the actual *unique key integrity* error it was much easier to understand what was happening.

```
public function __construct(
    \Magento\Catalog\Api\CategoryLinkManagementInterface $categoryLinkManagementInterface
) {
    $this->categoryLinkManagementInterface = $categoryLinkManagementInterface;
}

/**
 * Assign product to category
 * @param \Magento\Catalog\Model\Product $product
 * @param array $categoryIds
 * @return void
 */
public function assignProductToCategory($product, $categoryIds = [])
{
    if(!empty($categoryIds)) {
        $this->categoryLinkManagementInterface->assignProductToCategories(
            $product->getSku(),
            $categoryIds
        );
    }
```

### Contribution checklist (*)
 - [*] Pull request has a meaningful description of its purpose
 - [*] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=22514) <br/>All committers have signed the CLA.

**Review by orlangur — CHANGES_REQUESTED** [review]
Low-level exception messages must never be exposed to end-user. Add logging of such exception instead.

**Comment by DominicWatts** [comment]
Made suggested change

**Comment by magento-engcom-team** [comment]
@DominicWatts thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by p-bystritsky** [comment]
@sivaschenko requested changes done.

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php [hunk 1]
```diff
@@ -21,20 +26,28 @@ class CategoryLinkRepository implements \Magento\Catalog\Api\CategoryLinkReposit
...
     /**
      * @param \Magento\Catalog\Api\CategoryRepositoryInterface $categoryRepository
      * @param \Magento\Catalog\Api\ProductRepositoryInterface $productRepository
+     * @param LoggerInterface|null $logger
      */
     public function __construct(
         \Magento\Catalog\Api\CategoryRepositoryInterface $categoryRepository,
-        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository
+        \Magento\Catalog\Api\ProductRepositoryInterface $productRepository,
+        ?LoggerInterface $logger
```
Comment (orlangur) [inline app/code/Magento/Catalog/Model/CategoryLinkRepository.php]: Must not be nullable and should be `null` by default.

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php [hunk 2]
```diff
@@ -87,6 +101,7 @@ public function deleteByIds($categoryId, $sku)
         try {
             $category->save();
         } catch (\Exception $e) {
+            $this->logger->critical($e->getMessage());
```
Comment (orlangur) [inline app/code/Magento/Catalog/Model/CategoryLinkRepository.php:104]: Don't we have a separate method to log exception?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php (hunk 1 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Model/CategoryLinkRepository.php]: @orlangur is using nullable types prohibited?

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php (hunk 2 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Model/CategoryLinkRepository.php:104]: @orlangur seems like we don't.

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php (hunk 1 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Model/CategoryLinkRepository.php]: @p-bystritsky this is just not needed in constructors as there is no sense to pass `null` explicitly. You either use default `null` value or pass a correct object.

File: app/code/Magento/Catalog/Model/CategoryLinkRepository.php (hunk 2 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Model/CategoryLinkRepository.php:104]: Ok, will try to grep later.

**Comment by sivaschenko** [comment]
Hi @akaplya what do you think about adding an optional constructor parameter to `CategoryLinkRepository` ?

**Comment by sidolov** [comment]
@DominicWatts , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**